src/hotspot/share/gc/serial/serialArguments.cpp

```diff
@@ -23,10 +23,16 @@
  */
 
 #include "precompiled.hpp"
+#include "gc/shared/gcForwarding.hpp"
 #include "gc/shared/genArguments.hpp"
 #include "gc/serial/serialArguments.hpp"
 #include "gc/serial/serialHeap.hpp"
 
+void SerialArguments::initialize_heap_flags_and_sizes() {
+  GenArguments::initialize_heap_flags_and_sizes();
+  GCForwarding::initialize_flags(MaxNewSize + MaxOldSize);
```

Comment:
Good catch. This is actually a bug that is causing the CDS tests to fail. The used variables have not yet been initialized at this point. I tried making the suggested change and that fixed at least one of the CDS failures.

Reply:
Yes, one must, since MaxNewSize and MaxOldSize are still on their initial values, so way too large to allow the GC forwarding, and therefore CompactObjectHeaders get automatically disabled for SerialGC. That explains a bunch of the problems @lmesnik saw.

This fixes SerialGC for me:
```suggestion
  GCForwarding::initialize_flags(MaxHeapSize);
```